File: java/epi/MergeSortedLists.java
// Description:
//
// Merge two sorted linked lists into a single sorted list.

package epi;

import java.util.Arrays;
import java.util.List;
import rlib.ListUtils;
import rlib.LinkedListNode;
import static rlib.TestingUtils.expect;

public class MergeSortedLists {

  private static LinkedListNode<Integer> mergeTwoLists(
    LinkedListNode<Integer> headA, LinkedListNode<Integer> headB)
  {
    LinkedListNode<Integer> dummy = new LinkedListNode<>();
    LinkedListNode<Integer> curr = dummy;
    while(headA != null && headB != null) {
      if (headA.data < headB.data) {
        curr.next = headA;
        headA = headA.next;
      } else {
        curr.next = headB;
        headB = headB.next;
      }
      curr = curr.next;
    }
    curr.next = (headA == null) ? headB : headA;
    return dummy.next;
  }

  private static boolean check(
    LinkedListNode<Integer> head1, LinkedListNode<Integer> head2)
  {
    while (head1 != null || head2 != null) {
      if (head1.data != head2.data) return false;
      head1 = head1.next;
      head2 = head2.next;
    }
    if (head1 == null && head2 == null) return true;
    else return false;
  }

  private static void smallTest() {
    LinkedListNode<Integer> headA, headB, result, expected;

    headA = ListUtils.buildSingleList(Arrays.asList(1, 2, 3, 4, 5));
    headB = ListUtils.buildSingleList(Arrays.asList(6, 7, 8, 9, 10));
    result = mergeTwoLists(headA, headB);
    expected = ListUtils.buildSingleList(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9, 10));
    expect(check(result, expected)).toBe(true);

    headA = ListUtils.buildSingleList(Arrays.asList(2, 4, 6, 8, 10));
    headB = ListUtils.buildSingleList(Arrays.asList(1, 3, 5, 7, 9));
    result = mergeTwoLists(headA, headB);
    expected = ListUtils.buildSingleList(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9, 10));
  }

  public static void main(String[] args) {
    smallTest();
  }

}

File: python/epi/merge_sorted_lists.py
#!/usr/bin/env python3
#
# Description:
#
# Merge two sorted lists.

from rlib.linked_list_node import LinkedListNode
from rlib.testing_utils import expect

def merge_sorted_lists(L1, L2):
    dummy = tail = LinkedListNode()
    while L1 and L2:
        if L1.data < L2.data:
            tail.next = L1
            L1 = L1.next
        else:
            tail.next = L2
            L2 = L2.next
        tail = tail.next
    if L1: tail.next = L1
    else: tail.next = L2
    return dummy.next

def small_test():
    L1, L2 = None, None
    L = merge_sorted_lists(L1, L2)
    expect(L).to_be(None)
    L1 = LinkedListNode(123)
    L = merge_sorted_lists(L1, L2)
    expect(L.data).to_be(123)
    expect(L.next).to_be(None)
    L2 = LinkedListNode(123)
    L1 = None
    L = merge_sorted_lists(L1, L2)
    expect(L.data).to_be(123)
    expect(L.next).to_be(None)
    L1 = LinkedListNode(-123)
    L2 = LinkedListNode(123)
    L = merge_sorted_lists(L1, L2)
    expect(L.data).to_be(-123)
    expect(L.next.data).to_be(123)
    expect(L.next.next).to_be(None)

if __name__ == '__main__':
    small_test()
